### PR TITLE
implemented parser for Logd.0.log

### LINKED ISSUE #4061: Parser for logd.0.log from iOS' Sysdiagnose
Create a parser for the logd.0.log which can be found as part of an iOS Sysdiagnose dump. They can found at 

`<sysdiagnose_DATE_OS_DEVICE_BUILD>/system_logs.logarchive/system_logs.logarchive/extra/logs0.log`

### PR DESCRIPTION
**Pull request by rick-slin**
## One line description of pull request

Parser for the logd.0.log file obtained from iOS Sysdiagnose file dumps.

## Description:

**Related issue (if applicable):** fixes #4061

## Notes:
All contributions to Plaso undergo [code
review](https://github.com/log2timeline/l2tdocs/blob/main/process/Code%20review%20process.asciidoc). This makes sure
that the code has appropriate test coverage and conforms to the Plaso [style
guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
  - [x] Automated checks (Travis, Codecov, Codefactor )pass
  - [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [x] Reviewer assigned

### CODE REVIEW

**Comment by joachimmetz** [comment]
@rick-slin thx for the proposed changes, the CI tests are currently failing because the changes do not match the style guide https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html

**Comment by joachimmetz** [comment]
assigning @jleaniz for a first review

File: plaso/parsers/logd.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    super(LogdEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.logger = None
+    self.body = None
+
+
+class LogdParser(text_parser.PyparsingSingleLineTextParser):
+  """Parser for the logd.0.log file obtained from iOS Sysdiagnose file dumps."""
+
+  NAME = 'logd.0.log'
+  DATA_FORMAT = 'logd log'
```
Comment (jleaniz) [inline plaso/parsers/logd.py:37]: Minor comment but I'd suggest to mention this is for iOS sysdiagnose logs as part of the description. 
```
DATA_FORMAT = 'iOS sysdiagnose log file'
```

File: plaso/parsers/logd.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import errors
+from plaso.lib import definitions
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class LogdEventData(events.EventData):
```
Comment (jleaniz) [inline plaso/parsers/logd.py:16]: I suggest we change LogEventData to something more concrete like IOSysdiagnoseLogData or similar.

File: plaso/parsers/logd.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  """logd event data
+
+  Attributes:
+    logger (str): process that submits the log event
+    body (str): body of the event line
+  """
+
+  DATA_TYPE = "logd:line"
```
Comment (jleaniz) [inline plaso/parsers/logd.py:24]: suggestion: logd:line => ios_sysdiagnose_logd:line ?

**Review by jleaniz — CHANGES_REQUESTED** [review]
@rick-slin Thanks for the parser. I left some minor comments.

**Comment by jleaniz** [comment]
@Onager please take a look when time permits.

**Comment by joachimmetz** [comment]
@Onager is currently pre-occuplied

File: data/formatters/ios.yaml [hunk 4]
```diff
@@ -78,3 +78,11 @@ message:
...
 - 'Name: {name}'
 - 'Message: {text}'
+---
+type: 'conditional'
+data_type: 'logd:line'
+message:
+- 'Logger: {logger}'
+- 'Body: {body}'
+short_message:
+  - 'Body: {body}'
```
Comment (joachimmetz) [inline data/formatters/ios.yaml:88]: consistency nit: use the same indentation, I'll change this

File: data/formatters/ios.yaml [hunk 5]
```diff
@@ -78,3 +78,11 @@ message:
 short_message:
 - 'Name: {name}'
 - 'Message: {text}'
+---
+type: 'conditional'
+data_type: 'logd:line'
```
Comment (joachimmetz) [inline data/formatters/ios.yaml:83]: add `ios`

File: plaso/parsers/ios_logd.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+class IOSysdiagnoseLogdData(events.EventData):
+  """logd event data
+
+  Attributes:
+    logger (str): process that submits the log event
+    body (str): body of the event line
+  """
+
+  DATA_TYPE = "ios_sysdiagnose_logd:line"
```
[the remark this answers is not in the document]
Comment (joachimmetz) [reply plaso/parsers/ios_logd.py:24]: I also opt to use `ios:sysdiagnose:logd:line`

File: plaso/parsers/__init__.py [hunk 7]
```diff
@@ -29,6 +29,7 @@
 from plaso.parsers import iis
 from plaso.parsers import java_idx
 from plaso.parsers import locate
+from plaso.parsers import ios_logd
```
Comment (joachimmetz) [inline plaso/parsers/__init__.py:32]: style: nit use alphabetical order

File: plaso/parsers/ios_logd.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import errors
+from plaso.lib import definitions
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class IOSysdiagnoseLogdData(events.EventData):
+  """logd event data
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:17]: Style nit: consider docstrings as sentences

File: plaso/parsers/ios_logd.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+  Attributes:
+    logger (str): process that submits the log event
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:20]: style nit: alphabetical ordering

File: plaso/parsers/ios_logd.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    body (str): body of the event line
+  """
+
+  DATA_TYPE = "ios_sysdiagnose_logd:line"
+
+  def __init__(self):
+    """Initializes event data."""
+    super(IOSysdiagnoseLogdData, self).__init__(data_type=self.DATA_TYPE)
+    self.logger = None
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:29]: style nit: alphabetical ordering

File: plaso/parsers/ios_logd.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    self.body = None
+
+
+class IOSysdiagnoseLogdParser(text_parser.PyparsingSingleLineTextParser):
+  """Parser for the logd.0.log file obtained from iOS Sysdiagnose file dumps."""
+
+  NAME = 'logd.0.log'
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:36]: Note that this is the name of the parser not the file

File: plaso/parsers/ios_logd.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+  DATE_ELEMENTS = text_parser.PyparsingConstants.DATE_ELEMENTS
+  TIME_ELEMENTS = text_parser.PyparsingConstants.TIME_ELEMENTS
+
+  _TIMESTAMP = DATE_ELEMENTS + TIME_ELEMENTS
+
+  _TIME_DELTA = pyparsing.Word(
+      pyparsing.nums + '+' + '-', exact=5)('time_delta')
+
+  _LOGGER = pyparsing.SkipTo(':')('logger') + pyparsing.Suppress(': ')
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:47]: Use the long form `setResultsName` which is much easier to read/understand

File: plaso/parsers/ios_logd.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+  _BODY = pyparsing.SkipTo(pyparsing.LineEnd())('body')
+
+  _LINE_GRAMMAR = _TIMESTAMP + _TIME_DELTA + _LOGGER + _BODY
+
+  LINE_STRUCTURES = [('log_entry', _LINE_GRAMMAR)]
+
+  def ParseRecord(self, parser_mediator, key, structure):
+    """Parses a log record structure and produces events.
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:56]: style nit: add an empty line

File: plaso/parsers/ios_logd.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): tokens from a parsed log line.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+   """
+
+    if key != 'log_entry':
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:70]: style nit: indentation does not align with docstring

File: plaso/parsers/ios_logd.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+
+    year = self._GetValueFromStructure(structure, 'year')
+    month = self._GetValueFromStructure(structure, 'month')
+    day = self._GetValueFromStructure(structure, 'day_of_month')
+    hours = self._GetValueFromStructure(structure, 'hours')
+    minutes = self._GetValueFromStructure(structure, 'minutes')
+    seconds = self._GetValueFromStructure(structure, 'seconds')
+
+    time_delta = self._GetValueFromStructure(structure, 'time_delta')
+    time_deta_hours = time_delta[:3]
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:82]: I assume `deta` should be `delta`

File: plaso/parsers/ios_logd.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    # dfdatetime takes the time zone offset as an int number of minutes but
+    # inverts the sign for example Easter Standard Time, which is UTC-5:00 needs
+    # to be provided as +300 minutes.
+    time_offset = -1 * int(time_deta_hours) * 60 + int(time_delta[3:])
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:86]: use int(value, 10) and catch ValueError

File: plaso/parsers/ios_logd.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+      parser_mediator.ProduceExtractionWarning('invalid date time value')
+      return
+
+    event = time_events.DateTimeValuesEvent(
+        date_time, definitions.TIME_DESCRIPTION_MODIFICATION)
+
+    parser_mediator.ProduceEventWithEventData(event, event_data)
+
+  def VerifyStructure(self, parser_mediator, line):
+    """Verifies that this is a mobile installation log file.
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:106]: `mobile installation log file` copy paste error ?

File: plaso/parsers/ios_logd.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import errors
+from plaso.lib import definitions
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class IOSysdiagnoseLogdData(events.EventData):
```
Comment (joachimmetz) [inline plaso/parsers/ios_logd.py:16]: Missing S? `IOSSysdiagnoseLogdData`

File: tests/parsers/ios_logd.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""Tests for the iOS Mobile Installation log parser."""
```
Comment (joachimmetz) [inline tests/parsers/ios_logd.py:3]: `iOS Mobile Installation log parser` copy paste error?

File: tests/parsers/ios_logd.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+import unittest
+
+from plaso.parsers import ios_logd
+
+from tests.parsers import test_lib
+
+
+class IOSysdiagnoseLogdUnitTest(test_lib.ParserTestCase):
```
Comment (joachimmetz) [inline tests/parsers/ios_logd.py:12]: Missing S ?

File: tests/parsers/ios_logd.py [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    number_of_warnings = storage_writer.GetNumberOfAttributeContainers(
+        'extraction_warning')
+    self.assertEqual(number_of_warnings, 0)
+
+    events = list(storage_writer.GetEvents())
+
+    expected_event_values = {
+        'timestamp': '2021-08-10 22:50:23.000000',
+        'logger': 'logd[29]',
+        'body': 'libtrace_kic=1'
```
Comment (joachimmetz) [inline tests/parsers/ios_logd.py:32]: just put the } at the end of the last line

**Review by joachimmetz — COMMENTED** [review]
Numerous style nits, I'll make the changes

**Comment by joachimmetz** [comment]
@rick-slin @jleaniz left some comments mostly for educational purposes, made the changes already

**Review by joachimmetz — APPROVED** [review]
LGTM after latest changes

**Review by jleaniz — APPROVED** [review]

File: data/formatters/ios.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (joachimmetz) [reply data/formatters/ios.yaml]: says:
```
---
type: 'conditional'
data_type: 'ios:sysdiagnose:logd:line'
message:
- 'Logger: {logger}'
```

for me, are you looking at an outdated version?

[the remark this answers is not in the document]
Comment (jleaniz) [reply data/formatters/ios.yaml]: This was an old comment/change that had been requested. Closing this as resolved in your last commit :)

**Pull request merged**